### PR TITLE
Text Sets: Easier/smarter text font & color styling

### LINKED ISSUE #7995: Text sets: Easier/smarter text font & color styling
See [Figma](https://www.figma.com/file/1YM3jA9h4QGc4xLdyaXIJS/Stories-Sprint-1.7?node-id=8898%3A2153).

Note: this issue is for Text Sets specifically, please see the Text Sets part in Figma only, it's together with the Text Presets there.

## Description 

When adding text sets, we need to ensure the added text set will have enough colour contrast with its background.
Some or the ground work has been done for color calculation and image generation within #8118.

The following rules apply:
- If any of the texts in the set has a predefined color (anything else than the default black), ignore this text set and handle as it's currently being done (without any changes).
- The text elements within one text set have to get assigned the same color. E.g. we should not assign white color to one of the texts and black to the other within the same text set.
- If all the elements within one text set have at least AA level contrast using white color, all the elements will be assigned white color. If all the elements have AA contast using black color, all the elements will be assigned black color.
- If some of the elements need black color and some of the elements need white color, we will use a scrim behind the text set, depending on the better contrast, either 30% opacity white or 30% opacity black shape will be used to ensure the AA level for all elements. Note that for doing this we'll need to add a shape behind each relevant text set and either omit it upon insertion or use the scrim.
- If we're changing the colors of the texts, we should also change the colors of borders and shapes.

Note: Text Sets can also be dragged onto the page! To be confirmed if we should already change the color while dragging or only when dropping.
**^ Edit: for now, only when dropping.**

See [`contrastUtils`](https://github.com/google/web-stories-wp/blob/main/assets/src/edit-story/utils/contrastUtils.js) for the existing utils for contrast calculation.

## Examples
Screenshot 2021-07-02 at 15 13 57
Screenshot 2021-07-02 at 15 13 52

### PR DESCRIPTION
**Pull request by merapi**
## Summary

This PR ensures that any text set added to the page has good contrast with its background.
Uses either black/white color or semi-transparent black/white scrim (see [Figma](https://www.figma.com/file/1YM3jA9h4QGc4xLdyaXIJS/Stories-Sprint-1.7?node-id=8898%3A2153)).

Example:
image

## Relevant Technical Choices

[Firefox doesn't support](https://caniuse.com/?search=backdrop-filter) `backdrop-filter: blur()`, should we leave it as is or try to mimic that effect? The same applies for [text presets](https://github.com/google/web-stories-wp/pull/8118).

image

## To-do

- [ ] Tests

## User-facing changes

Text sets, when inserted (via click or drag) will change their colors based on the background.

### QA

Enable feature flag:
image

For easier testing, I've added a green dot on the text sets that are qualified for this feature because some of them are not obvious (in other words: exclude text sets with any predefined text color).

image

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

### UAT

- [ ] UAT should use the same steps as above.

This PR can be tested by following these steps:

1.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [ ] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [ ] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #7995

### CODE REVIEW

File: includes/Experiments.php [hunk 1]
```diff
@@ -355,14 +355,25 @@ public function get_experiments(): array {
...
 				'description' => __( 'Enable text insertion with smart color ensuring good contrast with the background', 'web-stories' ),
 				'group'       => 'editor',
 			],
+			/**
+			 * Author: @merapi
+			 * Issue #7995
+			 * Creation date: 2021-08-13
+			 */
+			[
+				'name'        => 'enableSmartTextSetsColor',
```
Comment (barklund) [inline includes/Experiments.php:360]: Is this flag actually used anywhere? Isn't the feature always enabled in the code currently?

File: packages/story-editor/src/app/canvas/canvasProvider.js [hunk 2]
```diff
@@ -47,6 +47,8 @@ function CanvasProvider({ children }) {
   const [eyedropperPixelData, setEyedropperPixelData] = useState(null);
   const [isEyedropperActive, setIsEyedropperActive] = useState(null);
   const [eyedropperCallback, setEyedropperCallback] = useState(null);
+  const [pageCanvasData, setPageCanvasData] = useState(null);
+  const [pageCanvasPromise, setPageCanvasPromise] = useState(null);
```
Comment (barklund) [inline packages/story-editor/src/app/canvas/canvasProvider.js:57]: Remember to mock these in the unit tests - there are a few failures due to undefined functions.

File: packages/story-editor/src/components/canvas/useInsertTextSet.js [hunk 3]
```diff
@@ -44,24 +49,142 @@ function useInsertTextSet() {
...
+              }
+              if (e.y < bb.y || !bb.y) {
+                bb.y = e.y;
+              }
+
+              return bb;
+            },
+            { x: null, y: null }
+          );
+          const { x, y } = textSetXY;
```
Comment (barklund) [inline packages/story-editor/src/components/canvas/useInsertTextSet.js]: You can just map the values out and pass the whole array to `Math.min`:

```suggestion
          const getMinDim = (dim) => Math.min.apply(null, toAdd.map(e=>e[dim]));
          const x = getMinDim('x');
          const y = getMinDim('y');
```

File: packages/story-editor/src/components/canvas/useInsertTextSet.js [hunk 4]
```diff
@@ -44,24 +49,142 @@ function useInsertTextSet() {
...
-      toAdd.forEach((element) => {
+
+      const hasPredefinedColor = toAdd.some((element) =>
+        element.content?.includes('color:')
+      );
+
+      let textElementsContrasts = [];
+      let firstValidColor = null;
+
+      const enableSmartTextColor = true;
```
Comment (barklund) [inline packages/story-editor/src/components/canvas/useInsertTextSet.js]: Ah, this should probably use the feature flag, I guess?

File: packages/story-editor/src/elements/shape/display.js [hunk 5]
```diff
@@ -42,6 +42,8 @@ const Element = styled.div`
   ${elementFillContent}
   ${elementWithBackgroundColor}
   ${elementWithBorder}
+  ${({ backdropBlur }) =>
+    backdropBlur && `backdrop-filter: blur(${backdropBlur}px)`}
```
Comment (barklund) [inline packages/story-editor/src/elements/shape/display.js:46]: Hm, was this specified anywhere or is it just because it looks good?

**Review by barklund — APPROVED** [review]
The feature works great (albeit a bit slow, but that was probably also the case before).

Just a few comments on the code, but overall it's a great experience! Needs heavy poking in bug-bash, but I can't seem to break it at least.

File: includes/Experiments.php (hunk 1 above)
Comment (merapi) [reply includes/Experiments.php:360]: Create a flag - good.
Actually use the flag - even better 😄 .
Thanks for catching that!

File: packages/story-editor/src/elements/shape/display.js (hunk 5 above)
Comment (merapi) [reply packages/story-editor/src/elements/shape/display.js:46]: Figma says so:

image

**Comment by barklund** [comment]
@googlebot I consent.

**Comment by swissspidy** [comment]
@barklund @miina With Marcin OOO & with QA passed, can one of you remove the green dots before merging?

**Comment by miina** [comment]
@googlebot I consent.

**Comment by merapi** [comment]
Thanks @miina. I've also added one simple test.

**Pull request merged**